lib/zone.ts

```diff
@@ -1232,5 +1253,151 @@ const Zone: ZoneType = (function(global: any) {
+    NativeError.stackTraceLimit = Math.max(NativeError.stackTraceLimit, 15);
+
+    // make sure that ZoneAwareError has the same property which forwards to NativeError.
+    Object.defineProperty(ZoneAwareError, 'stackTraceLimit', {
+      get: function() { return NativeError.stackTraceLimit; },
+      set: function(value) { return NativeError.stackTraceLimit = value; }
+    });
+  }
+
+  // Now we need to populet the `blacklistedStackFrames` as well as find the
+  // run/runGuraded/runTask frames. This is done by creating a detect zone and then threading
+  // the execution through all of the above methods so that we can look at the stack trace and
+  // find the frames of interest.
+  let detectZone: Zone = Zone.current.fork({
+    name: 'detect',
+    onInvoke: function(parentZoneDelegate: ZoneDelegate, currentZone: Zone, targetZone: Zone,
+                        delegate: Function, applyThis: any, applyArgs: any[], source: string): any {
+      // Here only so that it will show up in the stack frame so that it can be black listed.
+      return parentZoneDelegate.invoke(targetZone, delegate, applyThis, applyArgs, source);
+    },
+    onHandleError: function(parentZD: ZoneDelegate, current: Zone, target: Zone, error: any): boolean {
+      let frames = error.originalStack.split(/\n/);
+      let runFrame = false, runGuardedFrame = false, runTaskFrame = false;
+      while(frames.length) {
+        let frame = frames.shift();
+        // On safari it is possible to have stack frame with no line number.
+        // This check makes sure that we don't filter frames on name only (must have linenumber)
+        if (/:\d+:\d+/.test(frame)) {
+          // Get rid of the path so that we don't accidintely find function name in path.
+          // In chrome the seperator is `(` and `@` in FF and safari
+          // Chrome: at Zone.run (zone.js:100)
+          // Chrome: at Zone.run (http://localhost:9876/base/build/lib/zone.js:100:24)
+          // FireFox: Zone.prototype.run@http://localhost:9876/base/build/lib/zone.js:101:24
+          // Safari: run@http://localhost:9876/base/build/lib/zone.js:101:24
+          let fnName: string = frame.split('(')[0].split('@')[0];
+          let frameType = FrameType.trasition;
+          if (fnName.indexOf('ZoneAwareError') !== -1) {
+            zoneAwareFrame = frame;
+          }
+          if (fnName.indexOf('runGuarded') !== -1) {
+            runGuardedFrame = true;
+          } else if (fnName.indexOf('runTask') !== -1) {
+            runTaskFrame = true;
+          } else if (fnName.indexOf('run') !== -1) {
+            runFrame = true;
+          } else {
+            frameType = FrameType.blacklist;
+          }
+          blacklistedStackFrames[frame] = frameType;
+          // Once we find all of the frames we can stop looking.
+          if (runFrame && runGuardedFrame && runTaskFrame) {
+            ZoneAwareError['rewriteStack'] = true;
```

Comment:
Seems odd that we are setting a global here. It will never be reset to false. Consider the following three errors:
`A B A'`, where A and A' are essentially the same error triggered in the same place in the code.  And B is the one that triggers this onHandleError. 

Because of this global A and A' will be handled differently, but they should not.